### PR TITLE
Use utf-8 to decode filename

### PR DESCRIPTION
**Pull request by ken2812221**
See bitcoin/bitcoin#13869

Enable unicode support for leveldb on Windows

CI result for applying this change is available in bitcoin/bitcoin#13787

### CODE REVIEW

**Comment by maflcko** [comment]
This won't cause any issues when down/upgrading from 0.16 to 0.17 while keeping the datadir?

**Comment by ken2812221** [comment]
@MarcoFalke The user must change their config file to be utf-8 encoded if the file contains any non-ascii characters in it. Before 0.17, they are read as ansi-encoded characters on Windows. Except this, I can't see it could any issue here.

**Comment by ken2812221** [comment]
I'll add the caution in the description of bitcoin/bitcoin#13426

**Comment by ken2812221** [comment]
@MarcoFalke Do you mean applying this change to 0.16? Is so, it will not cause any problem because we don't define UNICODE in 0.16

File: util/env_win.cc [hunk 1]
```diff
@@ -201,26 +201,24 @@ class Win32Env : public Env
     virtual void SleepForMicroseconds(int micros);
 };
 
+#ifdef UNICODE
+#define CP_AUTO CP_UTF8
+#else
+#define CP_AUTO CP_ACP
+#endif
```
Comment (ken2812221) [inline util/env_win.cc]: Do we need to keep ANSI support?

**Comment by luke-jr** [comment]
This no longer checks for UNICODE, so @MarcoFalke 's question should be re-answered (and the PR description updated)

**Comment by ken2812221** [comment]
@luke-jr If we don't backport this change into older version of Bitcoin Core. Then there would have no problem because it's using utf-8 facet to imbue path on current master. I checked if UNICODE defined to make sure this change can be backported.

**Comment by luke-jr** [comment]
Why is it a problem to backport now?

**Comment by ken2812221** [comment]
Because it use MBCS encoding on boost::filesystem::path on Bitcoin Core <= 0.17. It changed to utf-8 by bitcoin/bitcoin#13877

**Comment by laanwj** [comment]
Needs rebase after #14

**Comment by ken2812221** [comment]
rebased

**Comment by ken2812221** [comment]
@sipsorcery @NicolasDorier @xaya Are you willing to review this and give me an ACK or a NACK.

**Comment by xaya** [comment]
@ken2812221 I don't know if it causes or solves any other issues but it definitely solved this issue here:
https://github.com/bitcoin/bitcoin/issues/15092

so ack for our issue

**Comment by sipsorcery** [comment]
utACK f8e797a058b7a3993314e985dfdff8124214ba99

**Comment by NicolasDorier** [comment]
Sorry I am kind of lost about understanding those encoding problem.
This code seems correct to me after reviewing it, but I lack good understanding of the problem.

I don't understand why there is methods for both `std::wstring` and `std::string` and why Bitcoin Core is using `std::string`. And I feel that trying to investigate will bring me down a very deep in the rabbit hole about the weirdness of encoding and legacy windows code, to which I am not too eager to jump in.

I see nothing wrong with this code, and wish it to be merged as it fix bitcoin/bitcoin#15092 though.

**Comment by ken2812221** [comment]
@laanwj Could this be merged?

**Pull request merged**